### ISSUE TITLE
openssl: doesn't build with enabled tests [Windows]

### ISSUE THREAD

**Issue by JMoVS** [Linuxbrew/homebrew-core#761]
# Please follow the general troubleshooting steps first:
- [x] Ran `brew update` and retried your prior step?
- [x] Ran `brew doctor`, fixed as many issues as possible and retried your prior step?
- [x] If you're seeing permission errors tried running `sudo chown -R $(whoami) $(brew --prefix)`?

OpenSSL doesn't build on windows subsystem for linux currently. It fails in the make depend area.

https://gist.github.com/6a18df98b469ce93b5f22867efb62e37

Any ideas how to solve this?

**Comment by sjackman** [comment]
Try rebuilding `makedepend` from source:
```
brew reinstall makedepend -s
```

**Comment by sjackman** [comment]
This issue may be the same issue as https://github.com/Linuxbrew/brew/issues/95
One hypothesis I have is that the Windows kernel is not able to load executables that have been fixed up by `patchelf`.

**Comment by JMoVS** [comment]
as I already startet, I did the following:

`````
brew reinstall makedepend -s
brew install openssl -sv
`````

it didn't compile, but it is a different error now:

**[waiting on brew gist-logs to finally finish so I can update this response]**

aside from that `brew test hello` results in:
`````
brew install hello
==> Downloading https://linuxbrew.bintray.com/bottles/hello-2.10.x86_64_linux.bottle.tar.gz
######################################################################## 100,0%
==> Pouring hello-2.10.x86_64_linux.bottle.tar.gz
🍺  /home/justin/.linuxbrew/Cellar/hello/2.10: 52 files, 231.3K
justin@PC69EP5:~$ brew test hello
Testing hello
==> /home/justin/.linuxbrew/Cellar/hello/2.10/bin/hello --greeting=brew
Last 15 lines from /home/justin/.cache/Homebrew/Logs/hello/01.hello:
2016-08-13 16:20:06 +0200

/home/justin/.linuxbrew/Cellar/hello/2.10/bin/hello
--greeting=brew

/home/justin/.linuxbrew/Cellar/hello/2.10/bin/hello: 1: /home/justin/.linuxbrew/Cellar/hello/2.10/bin/hello: ELF: not found
/home/justin/.linuxbrew/Cellar/hello/2.10/bin/hello: 2: /home/justin/.linuxbrew/Cellar/hello/2.10/bin/hello: @@�?@�?0Q�td�?�?8: not found
/home/justin/.linuxbrew/Cellar/hello/2.10/bin/hello: 3: /home/justin/.linuxbrew/Cellar/hello/2.10/bin/hello: Syntax error: "(" unexpected
Error: hello: failed
Failed executing: /home/justin/.linuxbrew/Cellar/hello/2.10/bin/hello --greeting=brew
/home/justin/.linuxbrew/Library/Homebrew/formula.rb:1508:in `block in system'
/home/justin/.linuxbrew/Library/Homebrew/formula.rb:1445:in `open'
/home/justin/.linuxbrew/Library/Homebrew/formula.rb:1445:in `system'
/home/justin/.linuxbrew/Library/Taps/homebrew/homebrew-core/Formula/hello.rb:26:in `block in <class:Hello>'
/home/justin/.linuxbrew/Library/Homebrew/formula.rb:1350:in `block in run_test'
/home/justin/.linuxbrew/Library/Homebrew/extend/fileutils.rb:14:in `block in mktemp'
/home/justin/.linuxbrew/Library/Homebrew/extend/fileutils.rb:75:in `block in run'
/home/justin/.linuxbrew/Library/Homebrew/extend/fileutils.rb:75:in `chdir'
/home/justin/.linuxbrew/Library/Homebrew/extend/fileutils.rb:75:in `run'
/home/justin/.linuxbrew/Library/Homebrew/extend/fileutils.rb:13:in `mktemp'
/home/justin/.linuxbrew/Library/Homebrew/formula.rb:1344:in `run_test'
/home/justin/.linuxbrew/Library/Homebrew/test.rb:33:in `block in <main>'
/usr/lib/ruby/1.9.1/timeout.rb:69:in `timeout'
/home/justin/.linuxbrew/Library/Homebrew/test.rb:32:in `<main>'
`````

as I already had it handy, here is the output of self-compiled makedepend:

`````
justin@myPC:~$ brew test makedepend

Testing makedepend
==> /home/justin/.linuxbrew/Cellar/makedepend/1.0.5/bin/makedepend

`````

and from bottled version:

`````
~$ brew test makedepend
Testing makedepend
==> /home/justin/.linuxbrew/Cellar/makedepend/1.0.5/bin/makedepend
Last 15 lines from /home/justin/.cache/Homebrew/Logs/makedepend/01.makedepend:
2016-08-13 16:23:54 +0200

/home/justin/.linuxbrew/Cellar/makedepend/1.0.5/bin/makedepend

/home/justin/.linuxbrew/Cellar/makedepend/1.0.5/bin/makedepend: 1: /home/justin/.linuxbrew/Cellar/makedepend/1.0.5/bin/makedepend: ELF: not found
/home/justin/.linuxbrew/Cellar/makedepend/1.0.5/bin/makedepend: 3: /home/justin/.linuxbrew/Cellar/makedepend/1.0.5/bin/makedepend: Syntax error: "(" unexpected
Error: makedepend: failed
Failed executing: /home/justin/.linuxbrew/Cellar/makedepend/1.0.5/bin/makedepend
/home/justin/.linuxbrew/Library/Homebrew/formula.rb:1508:in `block in system'
/home/justin/.linuxbrew/Library/Homebrew/formula.rb:1445:in `open'
/home/justin/.linuxbrew/Library/Homebrew/formula.rb:1445:in `system'
/home/justin/.linuxbrew/Library/Taps/homebrew/homebrew-core/Formula/makedepend.rb:52:in `block in <class:Makedepend>'
/home/justin/.linuxbrew/Library/Homebrew/formula.rb:1350:in `block in run_test'
/home/justin/.linuxbrew/Library/Homebrew/extend/fileutils.rb:14:in `block in mktemp'
/home/justin/.linuxbrew/Library/Homebrew/extend/fileutils.rb:75:in `block in run'
/home/justin/.linuxbrew/Library/Homebrew/extend/fileutils.rb:75:in `chdir'
/home/justin/.linuxbrew/Library/Homebrew/extend/fileutils.rb:75:in `run'
/home/justin/.linuxbrew/Library/Homebrew/extend/fileutils.rb:13:in `mktemp'
/home/justin/.linuxbrew/Library/Homebrew/formula.rb:1344:in `run_test'
/home/justin/.linuxbrew/Library/Homebrew/test.rb:33:in `block in <main>'
/usr/lib/ruby/1.9.1/timeout.rb:69:in `timeout'
/home/justin/.linuxbrew/Library/Homebrew/test.rb:32:in `<main>'
`````

**Comment by JMoVS** [comment]
also, building `mkvtoolnix` fails with:

`````
==> ./configure --disable-debug --prefix=/home/justin/.linuxbrew/Cellar/mkvtoolnix/9.3.1_1 --without-curl --with-boost=/home/justin/.linuxbrew/opt/boost --with-extra-includes=/home/justin/.linuxbrew/opt/libogg/include;/home/justin/.linuxbrew/opt/libvorbis/include;/home/justin/.linuxbrew/opt/libebml/include;/home/justin/.linuxbrew/opt/libmatroska/include --with-extra-libs=/home/justin/.linuxbrew/opt/libogg/lib;/home/justin/.linuxbrew/opt/libvorbis/lib;/home/justin/.linuxbrew/opt/libebml/lib;/home/justin/.linuxbrew/opt/libmatroska/lib --disable-qt
Failed to execute: ./configure

`````
Is that related? Should I essentially start rebuilding all packages from source in homebrew? ;-)

https://gist.github.com/d1ea8e5e65483a65d41038748f206052

**Comment by sjackman** [comment]
Interesting that `brew install hello` doesn't work for you. I believe Linuxbrew/brew#95 stated that `brew install hello` did work for them. I've asked them to reconfirm.

> Should I essentially start rebuilding all packages from source in homebrew? ;-)

If bottles are not working on Windows (as it seems), always build from source using:
```sh
export HOMEBREW_BUILD_FROM_SOURCE=1
```

**Comment by sjackman** [comment]
> also, building mkvtoolnix fails with:
`Failed to execute: ./configure`

Not sure what's going on here. Try
```sh
brew install -dv mkvtoolnix
```
Then select shell from the interactive menu, and try executing `./configure`.

**Comment by JMoVS** [comment]
It's also me who reported https://github.com/Linuxbrew/brew/issues/95 - well, installing worked, but I didn't run it (I oversaw this or maybe was faster than you could edit your comment) ;-)

**Comment by JMoVS** [comment]
BTW: I updated the comment with the gist from the openssl installation attempt

**Comment by JMoVS** [comment]
`mkvtoolnix` issue was due to xz not working but failing silently, so there was no extraction happening. I rebuilt xz from source and mkvtoolnix is hitting some other errors now, more on that later

**Comment by sjackman** [comment]
Thanks for the clarification regarding `hello`. Progress. Good stuff.

**Comment by JMoVS** [comment]
I now rm-ed all packages and am currently reinstalling with homebrew_build_from_source set to 1 so I will see into what errors I'll run...

**Comment by JMoVS** [comment]
could you have a look at https://gist.github.com/63a03fdb21c0432664c592b284d67745 ? This is the next attempt with all dependencies built from source

**Comment by JMoVS** [comment]
hmm, does this indicate that make depend shouldn't be run anymore? https://groups.google.com/forum/#!topic/mailing.openssl.dev/ePwMzh74GOE

**Comment by sjackman** [comment]
Try
```sh
brew install openssl --without-test
```
See
```sh
brew options openssl
```

**Comment by JMoVS** [comment]
ok, this worked, cool!

**Comment by JMoVS** [comment]
hmm, is this "solution" a thing that could be incorporated into linux/win formulas to account for this executable stack nonsense, @sjackman ?
https://github.com/Microsoft/BashOnWindows/issues/286#issuecomment-248125908

**Comment by sjackman** [comment]
Possibly, yes. For installed programs, it looks as though `execstack -c $HOMEBREW_PREFIX/bin/*` would workaround the issue by enabling the executable stack on all programs. It defeats the intended security benefit of disabling executable stack. There's probably a limited number of programs that need an executable stack. We could fix this on a case-by-case basis.

For the OpenSSL test it's trickier. I think that's an upstream bug that needs to be raised with the upstream OpenSSL developers.

**Comment by JMoVS** [comment]
hmm, starting with this formula which requires executable stack as well, how could I include the execstack command into the formula (also, execstack seems to require root, so probably *interesting* to include it into a brew formula) of openssl so I can build with tests? Would it just go at a specific position in the formula?

**Comment by sjackman** [comment]
I believe you'd have to somehow tell OpenSSL to use `execstack` when it runs its tests. You'd need help from the developers of OpenSSL for that.

**Comment by JMoVS** [comment]
Hmm, I was wondering about this executable stack as it seems that sometimes, this is already needed at brew install time. But I'm not sure.

**Comment by yorickdowne** [comment]
OpenSSL does not require execstack. Debian worked around the issue with -Wa,--noexecstack in ASFLAGS.  See https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=430583
A patch to have OpenSSL not request execstack was rejected in 2005, for portability issues. See http://openssl.6102.n7.nabble.com/PATCH-libcrypto-without-executable-stack-td30772.html 
I have an idea on how to create a portable patch, and may suggest that to the OpenSSL devs.
In the meantime, though, adding -Wa,--noexecstack to your build recipe if you find yourself on ELF using GCC/GAS and either on Linux, FreeBSD or Android, might be a good idea.

**Comment by sjackman** [comment]
Without looking at the `openssl` code, it's likely a bug in their Makefile that when `brew` sets `CFLAGS` it overrides `-Wa,--noexecstack`, which is not desirable behaviour of a Makefile. I've merged PR #1510 which should resolve this issue.
